File: impls/c/step1_read_print.c
#include <stdio.h>
#include <stdlib.h>
#include "reader.h"

#define PROMPT printf("user> ")

char *read(char *input)
{
    return input;
}

char *eval(char *sexp)
{
    return sexp;
}

char *print(char *value)
{
    return value;
}

void rep(char *input)
{
    printf("%s", print(eval(read(input))));
}

int main()
{
    char *linep = NULL;
    size_t linecap = 0;
    PROMPT;
    int ret = getline(&linep, &linecap, stdin);
    while (ret > -1) {
        rep(linep);
        PROMPT;
        ret = getline(&linep, &linecap, stdin);
    }
}

File: impls/zig/step1_read_print.zig
const std = @import("std");

const readLine = @import("util.zig").readLine;

const List = struct {
    first: LispValue, rest: List
};

const LispValue = union {
    int: i64, float: f64, bool: bool, list: *List
};

// fn recursiveDescent(alloc: *Allocator, buf: []u8) !?*LispValue {
//     switch (buf[0]) {
//     };
// }

fn READ(a: []u8) []u8 {
    return a;
}

fn EVAL(a: []u8) []u8 {
    return a;
}

fn PRINT(a: []u8) []u8 {
    return a;
}

fn rep(input: []u8) []u8 {
    var read_input = READ(input);
    var eval_input = EVAL(read_input);
    var print_input = PRINT(eval_input);
    return print_input;
}

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();
    const stdin = std.io.getStdIn().reader();
    while (true) {
        var buf: [1000]u8 = undefined;
        try stdout.print("user=> ", .{});
        if (try readLine(stdin, buf[0..])) |line| {
            try stdout.print("{s}\n", .{rep(line)});
        } else {
            break;
        }
    }
}
